// pomelo-rpc/lib/rpc-server/acceptors/tcp-acceptor.js
var EventEmitter = require('events').EventEmitter;
var util = require('util');
var utils = require('../../util/utils');
var net = require('net');
var Composer = require('stream-pkg');
var logger = require('pomelo-logger').getLogger('pomelo-rpc', __filename);
var Tracer = require('../../util/tracer');

var Acceptor = function (opts, cb) {
    EventEmitter.call(this);
    opts = opts || {};
    this.bufferMsg = opts.bufferMsg;
    this.debugBufferMsgInterval = opts.debugBufferMsgInterval;
    this.interval = opts.interval;  // flush interval in ms
    this.pkgSize = opts.pkgSize;
    this.rpcDebugLog = opts.rpcDebugLog;
    this.rpcLogger = opts.rpcLogger;
    this.whitelist = opts.whitelist;
    this._interval = null;          // interval object
    this.server = null;
    this.sockets = {};
    this.msgQueues = {};
    this.cb = cb;
};
util.inherits(Acceptor, EventEmitter);

var pro = Acceptor.prototype;

pro.listen = function (port) {
    //check status
    if (!!this.inited) {
        utils.invokeCallback(this.cb, new Error('already inited.'));
        return;
    }
    this.inited = true;

    var self = this;

    this.server = net.createServer();
    this.server.listen(port);

    this.server.on('error', function (err) {
        logger.error('rpc server port: %s is error: %s', port, err.stack);
        self.emit('error', err);
    });

    this.server.on('connection', function (socket) {
        socket.id = '' + socket.remoteAddress + ':' + socket.remotePort;
        self.sockets[socket.id] = socket;
        socket.composer = new Composer({maxLength: self.pkgSize});

        logger.info('rpc server opened connection with remote client id: %j, address: %j',
                    socket.id, socket.address());

        self.emit('connection', {id: socket.id, ip: socket.remoteAddress});


        socket.on('data', function (data) {
            socket.composer.feed(data);
        });

        socket.composer.on('data', function (data) {
            try {
                var pkg = JSON.parse(data.toString());
                if (pkg instanceof Array) {
                    processMsgs(socket, self, pkg);
                } else {
                    processMsg(socket, self, pkg);
                }
            } catch (e) {
                // socke.io would broken if uncaugth the exception
                logger.error('rpc server process message error: %s', e.stack);
            }
        });

        socket.on('close', function (had_error) {
            logger.error('rpc server disconnect with remote client had_error: %s, remote: %j', had_error,
                         {id: socket.id, address: socket.remoteAddress});
            delete self.sockets[socket.id];
            delete self.msgQueues[socket.id];
        });

        socket.on('error', function (err) {
            logger.error('rpc server socket port: %s is error: %s', port, err.stack);
            self.emit('error', err);
        });

    });

    this.on('connection', ipFilter.bind(this));

    if (this.bufferMsg) {
        this._interval = setInterval(function () {
            flush(self);
        }, this.interval);
    }

    if (!!this.debugBufferMsgInterval) {
        setInterval(function () {
            var message = '';
            var total = 0;
            for (k in self.msgQueues) {
                total += self.msgQueues[k].length;
                if (!self.sockets[k]) {
                    message += util.format('[no socket]:%s ', self.msgQueues[k].length);
                }
                else {
                    message += util.format('[%j]:%s ', self.sockets[k].id, self.msgQueues[k].length);
                }
            }
            if( total > 0 ) {
                logger.info('msgQueues size bufferMsg: %j, total: %s, %s', self.bufferMsg, total, message);
            }
        }, this.debugBufferMsgInterval);
    }
};

var ipFilter = function (obj) {
    if (typeof this.whitelist === 'function') {
        var self = this;
        self.whitelist(function (err, tmpList) {
            if (err) {
                logger.error('%j.(RPC whitelist).', err);
                return;
            }
            if (!Array.isArray(tmpList)) {
                logger.error('%j is not an array.(RPC whitelist).', tmpList);
                return;
            }
            if (!!obj && !!obj.ip && !!obj.id) {
                for (var i in tmpList) {
                    var exp = new RegExp(tmpList[i]);
                    if (exp.test(obj.ip)) {
                        return;
                    }
                }
                var sock = self.sockets[obj.id];
                if (sock) {
                    sock.disconnect('unauthorized');
                    logger.warn('%s is rejected(RPC whitelist).', obj.ip);
                }
            }
        });
    }
};

pro.close = function () {
    if (!!this.closed) {
        return;
    }
    this.closed = true;
    if (this._interval) {
        clearInterval(this._interval);
        this._interval = null;
    }
    try {
        this.server.close();
    } catch (err) {
        logger.error('rpc server close error: %j', err.stack);
    }
    this.emit('closed');
};

var cloneError = function (origin) {
    // copy the stack infos for Error instance json result is empty
    var res = {
        msg: origin.msg,
        stack: origin.stack
    };
    return res;
};

var processMsg = function (socket, acceptor, pkg) {
    var tracer = new Tracer(acceptor.rpcLogger, acceptor.rpcDebugLog, pkg.remote, pkg.source, pkg.msg, pkg.traceId,
                            pkg.seqId);
    tracer.info('server', __filename, 'processMsg', 'tcp-acceptor receive message and try to process message');
    acceptor.cb.call(null, tracer, pkg.msg, function () {
        var args = Array.prototype.slice.call(arguments, 0);
        for (var i = 0, l = args.length; i < l; i++) {
            if (args[i] instanceof Error) {
                args[i] = cloneError(args[i]);
            }
        }
        var resp;
        if (tracer.isEnabled) {
            resp = {
                traceId: tracer.id,
                seqId: tracer.seq,
                source: tracer.source,
                id: pkg.id,
                resp: Array.prototype.slice.call(args, 0)
            };
        }
        else {
            resp = {id: pkg.id, resp: Array.prototype.slice.call(args, 0)};
        }
        if (acceptor.bufferMsg) {
            enqueue(socket, acceptor, resp);
        } else {
            socket.write(socket.composer.compose(JSON.stringify(resp)));
        }
    });
};

var processMsgs = function (socket, acceptor, pkgs) {
    for (var i = 0, l = pkgs.length; i < l; i++) {
        processMsg(socket, acceptor, pkgs[i]);
    }
};

var enqueue = function (socket, acceptor, msg) {
    var queue = acceptor.msgQueues[socket.id];
    if (!queue) {
        queue = acceptor.msgQueues[socket.id] = [];
    }
    queue.push(msg);
};

var flush = function (acceptor) {
    var sockets = acceptor.sockets, queues = acceptor.msgQueues, queue, socket;
    for (var socketId in queues) {
        socket = sockets[socketId];
        if (!socket) {
            // clear pending messages if the socket not exist any more
            delete queues[socketId];
            continue;
        }
        queue = queues[socketId];
        if (!queue.length) {
            continue;
        }
        socket.write(socket.composer.compose(JSON.stringify(queue)));
        queues[socketId] = [];
    }
};

/**
 * create acceptor
 *
 * @param opts init params
 * @param cb(tracer, msg, cb) callback function that would be invoked when new message arrives
 */
module.exports.create = function (opts, cb) {
    return new Acceptor(opts || {}, cb);
};

process.on('SIGINT', function () {
    process.exit();
});
